File: 2246-longest-path-with-different-adjacent-characters/2246-longest-path-with-different-adjacent-characters.py
class Solution(object):
    ans = 1
    def longestPath(self, parent, s):
        childs = collections.defaultdict(list)
        for i in range(len(parent)):
            childs[parent[i]].append(i)


        def dfs(node):
            if node not in childs:
                return 1

            path1 = 0
            for x in childs[node]:
                path2 = dfs(x)
                if s[x]!=s[node]:
                    self.ans = max(self.ans, path1 + path2 +1)

                    path1 = max(path1, path2)

            return path1+1

        dfs(0)
        return self.ans

File: 2246-longest-path-with-different-adjacent-characters/2246-longest-path-with-different-adjacent-characters.cpp
class Solution {
public:
    int help(vector<vector<int>>&chids, string &s, int n, int &res){
        if(!chids[n].size()) //leaf node
            return 1;
        int temp,path1=0,path2=0;
        for(auto &x: chids[n]){
            temp=help(chids,s,x,res);
            if(s[n]==s[x])
                continue;
            if(temp>path1){
                path2=path1;
                path1=temp;
            }
            else if(temp>path2)
                path2=temp;
        }
        res=max(res, 1+path1+path2);
        return 1+path1;
    }
    int longestPath(vector<int>& parent, string s) {
        int res=1,n=parent.size();
        vector<vector<int>>chids(n);
        for(int i=0;i<n;i++)
            if(parent[i]>=0)
                chids[parent[i]].push_back(i);
        help(chids,s,0,res);
        return res;
    }
};